src/append/rolling_file/policy/compound/trigger/time.rs

```diff
@@ -0,0 +1,135 @@
+//! The time trigger.
+//!
+//! Requires the `time_trigger` feature.
+
+#[cfg(feature = "file")]
+use serde::de;
+#[cfg(feature = "file")]
+use serde_derive::Deserialize;
+use std::error::Error;
+#[cfg(feature = "file")]
+use std::fmt;
+use std::sync::RwLock;
+
+use crate::append::rolling_file::policy::compound::now_string;
+use crate::append::rolling_file::policy::compound::trigger::Trigger;
+use crate::append::rolling_file::LogFile;
+
+#[cfg(feature = "file")]
+use crate::file::{Deserialize, Deserializers};
+
+/// Configuration for the time trigger.
+#[cfg(feature = "file")]
+#[derive(Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct TimeTriggerConfig {
+    #[serde(deserialize_with = "deserialize_unit")]
+    unit: String,
+}
+
+#[cfg(feature = "file")]
+fn deserialize_unit<'de, D>(d: D) -> Result<String, D::Error>
+where
+    D: de::Deserializer<'de>,
+{
+    struct V;
+
+    impl<'de2> de::Visitor<'de2> for V {
+        type Value = String;
+
+        fn expecting(&self, unit: &mut fmt::Formatter) -> fmt::Result {
+            unit.write_str("a time unit")
+        }
+
+        fn visit_str<E>(self, v: &str) -> Result<String, E>
+        where
+            E: de::Error,
+        {
+            match v {
+                "year" => Ok("%Y".to_owned()),
+                "month" => Ok("%Y%m".to_owned()),
+                "day" => Ok("%Y%m%d".to_owned()),
+                "hour" => Ok("%Y%m%d%H".to_owned()),
+                "minute" => Ok("%Y%m%d%H%M".to_owned()),
+                "second" => Ok("%Y%m%d%H%M%S".to_owned()),
+                _ => Err(E::invalid_value(
+                    de::Unexpected::Str(v),
+                    &"invalid unit, should be one of year, month, day, hour, minute, second",
+                )),
+            }
+        }
+    }
+
+    d.deserialize_any(V)
+}
+
+/// A trigger which rolls the log once it has passed a certain time.
+#[derive(Debug)]
+pub struct TimeTrigger {
+    fmt: String,
+    time_string: RwLock<String>,
```

Comment:
Why are we using an rwlock here?

Reply:
This time_string is used to compare whether to trigger or not. When triggered, time_string should be set to the current trigger time. It is not good to change from ` fn trigger(& self` to ` fn trigger(&mut self`, so use `Rwlock` can make the mutation.